File: section2/118.杨辉三角.py
#
# @lc app=leetcode.cn id=118 lang=python3
#
# [118] 杨辉三角
#

# @lc code=start
from typing import List


class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        res = []
        for index in range(1, numRows + 1):
            row = [1] * index
            for column in range(1, index - 1):
                row[column] = res[-1][column - 1] + res[-1][column]

            res.append(row)

        return res

    def test(self):
        print(self.generate(5) == [[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1]])
        print(self.generate(1) == [[1]])


Solution().test()

# @lc code=end

File: section2/section/118.杨辉三角.go
package section

import "fmt"

/*
 * @lc app=leetcode.cn id=118 lang=golang
 *
 * [118] 杨辉三角
 */

// @lc code=start
func generate(numRows int) [][]int {
	if numRows == 0 {
		return nil
	}

	rows := make([][]int, 0, numRows)
	rows = append(rows, []int{1})
	for length := 1; length < numRows; length++ {
		row := make([]int, 0, length+1)
		for i := 0; i <= length; i++ {
			column := 0
			if i-1 >= 0 {
				column += rows[length-1][i-1]
			}

			if i < length {
				column += rows[length-1][i]
			}
			row = append(row, column)
		}

		rows = append(rows, row)
	}

	return rows
}

func Test118() {
	// [[1],[1,1],[1,2,1],[1,3,3,1],[1,4,6,4,1]]
	fmt.Println(generate(5))
	// [[1]]
	fmt.Println(generate(1))
}

// @lc code=end
